src/agentscope/agents/rag_agents.py

```diff
@@ -0,0 +1,181 @@
+        self.retriever_list = []
+        self.description = kwargs.get("description", "")
+        self.rag_config = rag_config or {}
+
+    def reply(self, x: dict = None) -> dict:
+        """
+        Reply function of the RAG agent.
+        Processes the input data,
+        1) use the input data to retrieve with RAG function;
+        2) generates a prompt using the current memory and system
+        prompt;
+        3) invokes the language model to produce a response. The
+        response is then formatted and added to the dialogue memory.
+
+        Args:
+            x (`dict`, defaults to `None`):
+                A dictionary representing the user's input to the agent. This
+                input is added to the memory if provided. Defaults to
+                None.
+        Returns:
+            A dictionary representing the message generated by the agent in
+            response to the user's input.
+        """
+        retrieved_docs_to_string = ""
+        # record the input if needed
+        if self.memory:
+            self.memory.add(x)
+            # in case no input is provided (e.g., in msghub),
+            # use the memory as query
+            history = self.memory.get_memory(
+                recent_n=self.rag_config.get("recent_n_mem", 1),
+            )
+            query = (
+                "/n".join(
+                    [msg["content"] for msg in history],
+                )
+                if isinstance(history, list)
+                else str(history)
+            )
+        elif x is not None:
+            query = x["content"]
+        else:
+            query = ""
+
+        if len(query) > 0:
+            # when content has information, do retrieval
+            scores = []
+            for retriever in self.retriever_list:
+                retrieved_nodes = retriever.retrieve(str(query))
+                for node in retrieved_nodes:
+                    scores.append(node.score)
+                    retrieved_docs_to_string += (
+                        "\n>>>> score:"
+                        + str(node.score)
+                        + "\n>>>> source:"
+                        + str(node.node.get_metadata_str())
+                        + "\n>>>> content:"
+                        + node.get_content()
+                    )
+
+            if self.rag_config["log_retrieval"]:
+                self.speak("[retrieved]:" + retrieved_docs_to_string)
+
+            if max(scores) < 0.4:
+                # if the max score is lower than 0.4, then we let LLM
+                # decide whether the retrieved content is relevant
+                # to the user input.
+                msg = Msg(
+                    name="user",
+                    role="user",
+                    content=CHECKING_PROMPT.format(
+                        retrieved_docs_to_string,
+                        query,
+                    ),
+                )
+                checking = self.model([msg])
```

Comment:
should use `self.model.format` before calling model ?

Reply:
added